utils/node_local_test.py

```diff
@@ -42,20 +42,146 @@ class NLT_Conf():
+        entry['severity'] = 'ERROR'
+        self.issues.append(entry)
+        self.close()
+
+    def explain(self, line, log_file, esignal):
+        """Log an error, along with the other errors it caused
+
+        Log the line as an error, and reference everything in the pending
+        array.
+        """
+        count = len(self.pending)
+        symptoms = set()
+        locs = set()
+        mtype = 'Fault injection'
+
+        sev = 'LOW'
+        if esignal:
+            symptoms.add('Process died with signal {}'.format(esignal))
+            sev = 'ERROR'
+            mtype = 'Fault injection caused crash'
+            count += 1
+
+        if count == 0:
+            print('Nothing to explain')
+            return
+
+        for (sline, smessage) in self.pending:
+            locs.add('{}:{}'.format(sline.filename, sline.lineno))
+            symptoms.add(smessage)
+
+        preamble = 'Fault injected here caused {} errors,' \
+                   ' logfile {}:'.format(count, log_file)
+
+        message = '{} {} {}'.format(preamble,
+                                    ' '.join(sorted(symptoms)),
+                                    ' '.join(sorted(locs)))
+        self.add(line,
+                 sev,
+                 message,
+                 cat='Fault injection location',
+                 mtype=mtype)
+        self.pending = []
+
+    def add(self, line, sev, message, cat=None, mtype=None):
+        """Log an error
+
+        Describe an error and add it to the issues array.
+        Add it to the pending array, for later clarification
+        """
+        entry = {}
+        entry['directory'] = os.path.dirname(line.filename)
+        entry['fileName'] = os.path.basename(line.filename)
+        if mtype:
+            entry['type'] = mtype
+        else:
+            entry['type'] = message
+        if cat:
+            entry['category'] = cat
+        entry['lineStart'] = line.lineno
+        entry['description'] = message
+        entry['message'] = line.get_anon_msg()
+        entry['severity'] = sev
+        if line.function in self.FLAKY_FUNCTIONS and \
+           entry['severity'] == 'NORMAL':
+            entry['severity'] = 'LOW'
+        self.issues.append(entry)
+        self.pending.append((line, message))
+        self._flush()
+
+    def _flush(self):
+        """Write the current list to the json file
+
+        This is done just in case of crash
+        """
+        self._fd.seek(0)
+        self._fd.truncate(0)
+        data = {}
+        data['issues'] = list(self.issues)
+        if self._running:
+            # When the test is running insert an error in case of abnormal
+            # exit, so that crashes in this code can be identified.
+            entry = {}
+            entry['fileName'] = os.path.basename(__file__)
+            entry['directory'] = os.path.dirname(__file__)
+            entry['lineStart'] = sys._getframe().f_lineno
```

Comment:
(pylint-protected-access) Access to a protected member _getframe of a client class